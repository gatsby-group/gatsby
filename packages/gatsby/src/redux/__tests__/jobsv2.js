const jobsManager = require(`../../utils/jobs-manager`)
jest.spyOn(jobsManager, `enqueueJob`)
jest.mock(`uuid/v4`, () => () => `1234`)

const { actions, internalActions } = require(`../actions`)
const jobsReducer = require(`../reducers/jobsv2`)

describe(`Job v2 actions/reducer`, () => {
  const plugin = {
    name: `test-plugin`,
    version: `1.0.0`,
  }
  const createGlobalState = defaultState => {
    return {
      program: {
        directory: __dirname,
      },
      jobsV2: jobsReducer(undefined, {}),
      ...defaultState,
    }
  }

  beforeEach(() => {
    jobsManager.enqueueJob.mockClear()
    jobsManager.enqueueJob.mockImplementation(() => Promise.resolve(`myresult`))
  })

  const createDispatcher = globalState => {
    const dispatch = realAction => jobsReducer(getState().jobsV2, realAction)
    const getState = () => globalState

    return action => action(dispatch, getState)
  }

  it(`should enqueueJob`, async () => {
    const globalState = createGlobalState()
    const dispatch = createDispatcher(globalState)

    const promise = dispatch(
      actions.createJobV2(
        {
          name: `TEST_JOB`,
          inputPaths: [],
          outputDir: `public/static`,
          args: {},
        },
        plugin
      )
    )

    expect(globalState.jobsV2).toMatchSnapshot()

    expect(globalState.jobsV2.done.size).toBe(0)
    expect(globalState.jobsV2.stale.size).toBe(1)
    expect(globalState.jobsV2.active.size).toBe(1)

    await promise

    const values = Array.from(globalState.jobsV2.done.values())
    expect(globalState.jobsV2.done.size).toBe(1)
    expect(values[0]).toEqual({
      result: `myresult`,
      inputPaths: [],
    })
    expect(jobsManager.enqueueJob).toMatchSnapshot()
  })

  it(`should return the result when job already ran`, async () => {
    const globalState = createGlobalState()
    const dispatch = createDispatcher(globalState)

    const job = {
      name: `TEST_JOB`,
      inputPaths: [],
      outputDir: `public/static`,
      args: {},
    }

    await expect(dispatch(actions.createJobV2(job, plugin))).resolves.toBe(
      `myresult`
    )
    await expect(dispatch(actions.createJobV2(job, plugin))).resolves.toBe(
      `myresult`
    )
    expect(jobsManager.enqueueJob).toHaveBeenCalledTimes(1)
  })

  it(`should remove a stale job`, async () => {
    const { jobsV2 } = createGlobalState()
    jobsV2.done.set(`1234`, {})
    jobsV2.stale.set(`12345`, {})

    jobsReducer(jobsV2, internalActions.removeStaleJob(`1234`))
    expect(jobsV2.done.size).toBe(0)
    expect(jobsV2.stale.size).toBe(1)

    jobsReducer(jobsV2, internalActions.removeStaleJob(`12345`))
    expect(jobsV2.done.size).toBe(0)
    expect(jobsV2.stale.size).toBe(0)
  })
})
