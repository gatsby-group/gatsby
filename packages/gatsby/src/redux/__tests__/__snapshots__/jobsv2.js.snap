// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Job v2 actions/reducer should enqueueJob 1`] = `
Object {
  "active": Set {
    "0264dcc4010e386cf4a9e94dc9e94af4",
  },
  "done": Map {},
  "stale": Map {
    "0264dcc4010e386cf4a9e94dc9e94af4" => Object {
      "job": Object {
        "args": Object {},
        "contentDigest": "0264dcc4010e386cf4a9e94dc9e94af4",
        "id": "1234",
        "inputPaths": Array [],
        "name": "TEST_JOB",
        "outputDir": "public/static",
        "plugin": Object {
          "name": "test-plugin",
          "version": "1.0.0",
        },
      },
      "plugin": Object {
        "name": "test-plugin",
        "version": "1.0.0",
      },
    },
  },
}
`;

exports[`Job v2 actions/reducer should enqueueJob 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "args": Object {},
        "contentDigest": "0264dcc4010e386cf4a9e94dc9e94af4",
        "id": "1234",
        "inputPaths": Array [],
        "name": "TEST_JOB",
        "outputDir": "public/static",
        "plugin": Object {
          "name": "test-plugin",
          "version": "1.0.0",
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;
